src/com/amazon/corretto/crypto/provider/Loader.java

```diff
@@ -289,68 +342,116 @@ private static void assertVersionMatch() {
         }
     }
 
+    private static void maybeDelete(final Path path) {
+        if (!DebugFlag.PRESERVE_NATIVE_LIBRARIES.isEnabled()) {
+            try {
+                Files.delete(path);
+            } catch (IOException ex) {
+                LOG.warning("Unable to delete native library: " + ex);
+            }
+        }
+    }
+
+    private static void maybeDeletePrivateTempDir(final Path tmpDirectory) {
+        for (String jarResource: JAR_RESOURCES) {
+            String resourceFileName = System.mapLibraryName(jarResource);
+            maybeDelete(tmpDirectory.resolve(resourceFileName));
+        }
+        maybeDelete(tmpDirectory);
+    }
+
+    private static Path writeJarResourceToTemporaryFile(Path tempDirectory, final String resourceFileName) throws IOException {
+        final Path tempResourceFilePath = tempDirectory.resolve(resourceFileName);
+
+        // Create a new temporary file. If one already exists this will throw FileAlreadyExistsException
+        Files.createFile(tempResourceFilePath, SELF_OWNER_FILE_PERMISSIONS);
+
+        try (InputStream is = Loader.class.getResourceAsStream(resourceFileName);
+             OutputStream os = Files.newOutputStream(tempResourceFilePath, StandardOpenOption.CREATE,
+                     StandardOpenOption.WRITE, StandardOpenOption.TRUNCATE_EXISTING)) {
+            final byte[] buffer = new byte[16 * 1024];
+            int read = is.read(buffer);
+            while (read >= 0) {
+                os.write(buffer, 0, read);
+                read = is.read(buffer);
+            }
+            os.flush();
+        }
+
+        // Ensure we delete any temp files on exit
+        tempResourceFilePath.toFile().deleteOnExit();
```

Comment:
here and in the directory creation routine, should we guard the `deleteOnExit()` call with `!DebugFlag.PRESERVE_NATIVE_LIBRARIES.isEnabled()`?

Reply:
No. We currently always want to delete on exit. That flag simply prevents immediate deletion (rather than deletion when the JVM terminates).

Reply:
Thanks! Added that check

Reply:
Done.